### PR TITLE
Enabling demos tests with nightly binaries

### PR DESCRIPTION
**Pull request by vedpatwardhan**
1. Added a way to dispatch the `.github/workflows/run-test-suite.yml` workflow with an input about whether binaries should be pulled from the `nightly` branch.
2. Mounting the volume of the latest clone of `ivy` everytime rather than the one present inside the docker image that gets updated once a day

### CODE REVIEW

**Comment by vedpatwardhan** [comment]
Hey @Dsantra92, here's an [example](https://github.com/unifyai/demos/actions/runs/6809112110) workflow with the nightly binaries, you'd need to go to the workflow page, select the branch as `demos_tests_nightly` (as we're yet to merge the PR) and pass "true" for the "Testing with nightly binaries : " field as shown in the image to trigger the workflow with the nightly binaries. Let me know if there's any changes to be made, thanks 😄
Screenshot 2023-11-09 at 2 12 53 PM

File: .github/workflows/_basic-test.yml [hunk 1]
```diff
@@ -1,5 +1,11 @@
 name: test-basics
-on: [workflow_call, workflow_dispatch]
+on: 
+  workflow_call:
+    inputs:
+      nightly_binaries:
+        type: string
+        default: "false"
+        required: false
```
Comment (Dsantra92) [inline .github/workflows/_basic-test.yml:8]: Are we removing the option for workflow_dispatch? Also can make the nightly_binaries type boolean, that seems more consistent?

File: .github/workflows/_demo-test.yml [hunk 2]
```diff
@@ -1,5 +1,11 @@
 name: test-examples
-on: [workflow_call, workflow_dispatch]
+on: 
+  workflow_call:
+    inputs:
+      nightly_binaries:
+        type: string
+        default: "false"
+        required: false
```
Comment (Dsantra92) [inline .github/workflows/_demo-test.yml:8]: Same as the review on .github/workflows/_basic-test.yml

File: .github/workflows/run-test-suite.yml [hunk 3]
```diff
@@ -1,6 +1,11 @@
 name: E2E-QA-and-Deployment
 on:
   workflow_dispatch:
+    inputs:
+      nightly_binaries:
+        description: 'Testing with nightly binaries :'
```
Comment (Dsantra92) [inline .github/workflows/run-test-suite.yml:6]: We should add the type here.

File: tests/test_demos.sh [hunk 4]
```diff
@@ -1,5 +1,10 @@
 #!/bin/bash
 
+if [ $3 == "true" ];
```
Comment (Dsantra92) [inline tests/test_demos.sh:3]: Sorry to nitpick on the formatting, lol:
```suggestion
if [ "$3" == "true" ]; then
    export VERSION=nightly
fi
```

**Review by Dsantra92 — CHANGES_REQUESTED** [review]
Some basic formatting changes. Otherwise LGTM.

File: .github/workflows/_basic-test.yml (hunk 1 above)
Comment (vedpatwardhan) [reply .github/workflows/_basic-test.yml:8]: actually the `workflow_dispatch` type is already present for the `run-test-suite` which calls these workflows anyway. Do you think we should have a `workflow_dispatch` for this workflow as well?
Sure, I'll change the type to `boolean` 👍

File: .github/workflows/run-test-suite.yml (hunk 3 above)
Comment (vedpatwardhan) [reply .github/workflows/run-test-suite.yml:6]: Sure, I'll add it 👍

File: tests/test_demos.sh (hunk 4 above)
Comment (vedpatwardhan) [reply tests/test_demos.sh:3]: no worries at all, it's actually better to format it that way 👍

Comment (vedpatwardhan) [reply tests/test_demos.sh]: done

File: .github/workflows/run-test-suite.yml (hunk 3 above)
Comment (vedpatwardhan) [reply .github/workflows/run-test-suite.yml:6]: done

File: .github/workflows/_demo-test.yml (hunk 2 above)
Comment (vedpatwardhan) [reply .github/workflows/_demo-test.yml]: done

File: .github/workflows/_basic-test.yml (hunk 1 above)
Comment (vedpatwardhan) [reply .github/workflows/_basic-test.yml:8]: done changing the type to `boolean`, let me know if you'd like a `workflow_dispatch` to `_basic-test`. Cuz for `_demos-test` we'd need `run-test-suite` as it needs to start the vm before running it and stop the vm after running `_demos-test`

Comment (Dsantra92) [reply .github/workflows/_basic-test.yml:8]: Yeah, I think that would too much of hassle for now. Moreover I don't think there will be cases where we need to run these seperately.

File: tests/test_demos.sh [hunk 5]
```diff
@@ -1,5 +1,9 @@
 #!/bin/bash
 
+if [ "$3" ]; then
```
Comment (Dsantra92) [inline tests/test_demos.sh:3]: This will still be a string comparison. The value of $3 is string as it passed to bash script.

**Review by Dsantra92 — CHANGES_REQUESTED** [review]

File: tests/test_demos.sh (hunk 5 above)
Comment (vedpatwardhan) [reply tests/test_demos.sh]: oh shoot, just made a change, let me know if that sounds good, thanks!

**Review by Dsantra92 — APPROVED** [review]

**Comment by Dsantra92** [comment]
LGTM. Feel free to merge when it's ready @vedpatwardhan.

**Comment by vedpatwardhan** [comment]
Awesome, thanks @Dsantra92 😄
Also, I saw a `nightly` branch on the `ivy` repo as well just like the `binaries` repo, do you want the demos tests running using the `nightly` branch binaries to also use the `ivy` repo's `nightly` branch?

**Comment by Dsantra92** [comment]
Yep, we would need the ivy nightly branch. Otherwise new binaries and function definitions wouldn't work.

**Comment by vedpatwardhan** [comment]
> Yep, we would need the ivy nightly branch. Otherwise new binaries and function definitions wouldn't work.

done, I'll merge the PR now, but there seems to be an additional test failure in the nightly branch with the [torch_to_jax](https://github.com/unifyai/demos/actions/runs/6864778182/job/18667351003) demo as compared to the current code and binaries. I'd suggest looking into that issue before merging the nightly binaries into main. Thanks @Dsantra92 😄

**Pull request merged**